src/odemis/driver/tmcm.py

```diff
@@ -1725,40 +1788,64 @@ def _doMoveAbs(self, future, pos):
             end = 0 # expected end
             old_pos = self._applyInversion(self.position.value)
             moving_axes = set()
+            moving_do_axes = set()
             for an, v in pos.items():
-                aid = self._name_to_axis[an]
-                moving_axes.add(aid)
-                usteps = int(round(v / self._ustepsize[aid]))
-                # Actual position is the one used for absolute move, so need to
-                # always reset it when there is an encoder
-                self._resetEncoderDeviation(aid, always=True)
-                self.MoveAbsPos(aid, usteps)
-                # compute expected end
-                try:
-                    d = abs(v - old_pos[an])
-                    dur = driver.estimateMoveDuration(d, self.speed.value[an], self._accel[an])
-                except Exception: # Can happen if config is wrong and report speed or accel == 0
-                    logging.exception("Failed to estimate move duration")
-                    dur = 60
-                end = max(time.time() + dur, end)
-
-            self._waitEndMove(future, moving_axes, end)
+                # Check if it's a digital output
+                if an in self._name_to_do_axis:
+                    port = self._name_to_do_axis[an]
+                    if self._leds_on and v is not self._led_prot_do[port]:
+                        # don't disable protection if lights are on
+                        logging.info("Ignoring request to change digital output on channel %s " % port +
```

Comment:
No, change the test to be `if self._leds_on and port in self._led_prot_do`. As soon as this condition applies, no change to the SetIO should ever be done. 
Inside, don't ignore the move, but only update `self._expected_do_pos`. 

You can still put the log only if `v is not self._led_prot_do[port]`, and update the log to just say that as referencing led is on, the DO axis move will be delayed.

Reply:
Regarding this: Please drop this code, and update the default value for `.referenced`:
What do you mean exactly? Isn't it doing the right thing if we replaced the second if with an elif?
```
if a in self._name_to_do_axis:
   continue  # no need to display warning for unreferenced do axis
elif not self.referenced.value.get(a, False) and p != self.position.value[a]:
   ...
```
This would avoid adding the do-axis to .referenced.

Reply:
line 1609, for some reason I cannot comment on your previous comment

Reply:
About the other comment (line 1797f): I still don't get it. 
```
    def _requestRefSwitch(self, axis):
        refswitch = self._refswitch.get(axis)
        if refswitch is None:
            return

        with self._refswitch_lock:
            # Set _leds_on attribute before closing shutters to make sure they are not
            # opened again in a concurrent thread
      ---->self._leds_on = True  # do this before closing shutters <--------------
            # Close shutters
            shutter_move = {}
            for channel, val in self._led_prot_do.items():
                do_an = self._do_axes[channel][0]
                # TODO: is this safe (are we sure .position is up to date)?
                if self.position.value[do_an] == val:
                    logging.debug("Shutters already closed, no need to close them again.")
                else:
                    self._expected_do_pos[do_an] = self.position.value[do_an]
                    shutter_move[do_an] = val
                
            if shutter_move:
                f = self._createMoveFuture()
     ------>  self._doMoveAbs(f, shutter_move) <------------------------------------------------
```
Here we turn the leds on and then we actually want to start moving the shutters (see my earlier comment why it's in that order).

If we test this `if self._leds_on and port in self._led_prot_do` in do_move_abs, it seems to me that we will not carry out that move and close the shutters.